pkg/chunk/local/boltdb_index_client.go

```diff
@@ -117,41 +119,45 @@ func (b *boltIndexClient) Stop() {
 	b.wait.Wait()
 }
 
-func (b *boltIndexClient) NewWriteBatch() chunk.WriteBatch {
+func (b *BoltIndexClient) NewWriteBatch() chunk.WriteBatch {
 	return &boltWriteBatch{
 		tables: map[string]map[string][]byte{},
 	}
 }
 
-func (b *boltIndexClient) getDB(name string) (*bbolt.DB, error) {
+func (b *BoltIndexClient) GetDB(name string, operation int) (*bbolt.DB, error) {
 	b.dbsMtx.RLock()
 	db, ok := b.dbs[name]
 	b.dbsMtx.RUnlock()
 	if ok {
 		return db, nil
 	}
 
-	b.dbsMtx.Lock()
-	defer b.dbsMtx.Unlock()
-	db, ok = b.dbs[name]
-	if ok {
-		return db, nil
-	}
+	if _, err := os.Stat(path.Join(b.cfg.Directory, name)); err == nil || operation == DBOperationWrite {
+		b.dbsMtx.Lock()
+		defer b.dbsMtx.Unlock()
+		db, ok = b.dbs[name]
+		if ok {
+			return db, nil
+		}
 
-	// Open the database.
-	// Set Timeout to avoid obtaining file lock wait indefinitely.
-	db, err := bbolt.Open(path.Join(b.cfg.Directory, name), 0666, &bbolt.Options{Timeout: 5 * time.Second})
-	if err != nil {
-		return nil, err
+		// Open the database.
+		// Set Timeout to avoid obtaining file lock wait indefinitely.
+		db, err := OpenBoltdbFile(path.Join(b.cfg.Directory, name))
```

Comment:
Could you explain the need of moving it to a dedicated function `OpenBoltdbFile()`? If it's required for an easier integration with Loki, could you explain how it will work, please?